java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.resolve;
+
+import com.google.common.base.Function;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Sets;
+
+import org.sonar.plugins.java.api.semantic.Type;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class TypeInferenceSolver {
+
+  private final Symbols symbols;
+
+  public TypeInferenceSolver(Symbols symbols) {
+    this.symbols = symbols;
+  }
+
+  TypeSubstitution inferTypeSubstitution(JavaSymbol.MethodJavaSymbol method, List<JavaType> formals, List<JavaType> argTypes) {
+    boolean isVarArgs = method.isVarArgs();
+    int numberFormals = formals.size();
+    int numberArgs = argTypes.size();
+    int numberParamToCheck = Math.min(numberFormals, numberArgs);
+    List<JavaType> newArgTypes = new ArrayList<>(argTypes);
+    TypeSubstitution substitution = new TypeSubstitution();
+
+    // method is varargs but parameter is not provided
+    if (isVarArgs && numberFormals == numberArgs + 1) {
+      numberParamToCheck += 1;
+      newArgTypes.add(symbols.objectType);
+    }
+    for (int i = 0; i < numberParamToCheck; i++) {
+      JavaType formalType = formals.get(i);
+      JavaType argType = newArgTypes.get(i);
+      boolean variableArity = isVarArgs && i == (numberFormals - 1);
+      List<JavaType> remainingArgTypes = new ArrayList<>(newArgTypes.subList(i, newArgTypes.size()));
+
+      substitution = inferTypeSubstitution(method, substitution, formalType, argType, variableArity, remainingArgTypes);
+
+      if (!method.isConstructor() && substitution.typeVariables().containsAll(method.typeVariableTypes)) {
+        // we found all the substitution
+        break;
+      }
+    }
+    return substitution;
+  }
+
+  private TypeSubstitution inferTypeSubstitution(JavaSymbol.MethodJavaSymbol method, TypeSubstitution currentSubstitution, JavaType formalType, JavaType argumentType,
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of returns of this method 6, down to the maximum allowed 5. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1142)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "inferTypeSubstitution" is 24 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)